whist/core/cards/card_container.py

```diff
@@ -1,60 +1,120 @@
+    _cards_set: set[Card] = PrivateAttr()
+
+    # pylint: disable=too-few-public-methods
+    class Config:
+        """
+        Configuration class for base model to make it immutable.
+        """
+        allow_mutation = False
+
+    def __init__(self, **data):
+        super().__init__(**data)
+        self._cards_set = set(self.cards)
+        self.__resync()
+
+    @classmethod
+    def empty(cls) -> 'CardContainer':
+        """
+        Creates an empty card container.
+
+        :return: empty ard container
+        :rtype: correct subtype of CardContainer
+        """
+        return cls(cards=())
+
+    @classmethod
+    def with_cards(cls, *cards) -> 'CardContainer':
+        """
+        Creates a card container with the given cards.
+
+        :param cards: cards to add
+        :return: card container with given cards
+        :rtype: correct subtype of CardContainer
+        """
+        if len(cards) == 1 and not isinstance(cards[0], Card):
+            cards = cards[0]
+        return cls(cards=cards)
+
+    @classmethod
+    def full(cls) -> 'CardContainer':
         """
-        Constructor
+        Create a full card container.
 
-        :param args: multiple cards or one card iterable
+        :return: full card container
+        :rtype: correct subtype of CardContainer
         """
-        if len(args) == 1 and not isinstance(args[0], Card):
-            self._cards = {*args[0]}
-        else:
-            self._cards = {*args}
+        return cls(cards=Card.all_cards())
+
+    def pop_random(self) -> Card:
+        """
+        Removes one random card from card container.
+
+        :return: A card from deck.
+        """
+        card = random.choice(self.cards)
+        self.remove(card)
+        return card
 
     def __contains__(self, card: Card) -> bool:
-        return card in self._cards
+        return card in self._cards_set
 
     def __len__(self):
-        return len(self._cards)
+        return len(self.cards)
 
     def __iter__(self) -> Iterator[Card]:
-        return iter(self._cards)
+        return iter(self.cards)
 
     def __str__(self) -> str:
-        return str(self._cards)
+        return str(self.cards)
 
     def __repr__(self) -> str:
-        return f'CardContainer(cards={self._cards!r})'
-
-    def __eq__(self, other: Any) -> bool:
-        if self.__class__ is other.__class__:
-            # pylint: disable=protected-access
-            return self._cards == other._cards
-        return NotImplemented
+        return f'CardContainer(cards={self.cards!r})'
 
     def remove(self, card: Card) -> None:
         """
         Remove a card from this container.
 
         :param card: card to remove
         """
-        self._cards.remove(card)
+        if not isinstance(card, Card):
+            raise ValueError(f'cannot remove {card} of type {type(card)} from card container')
+        if card not in self._cards_set:
+            raise ValueError(f'{card} not in card container')
```

Comment:
Maybe we should use customs errors here.

Reply:
ehh, no real need. I am using standard Python conventions here. Problems with arguments always raise `ValueError`

Reply:
Agreed, but it is much more convenient during exception handling to use custom errors. But we two different message here, I'm fine with it and it is up to you.

Reply:
ok